tini/src/seccomp_fd_notify.c

```diff
@@ -217,213 +226,130 @@ struct sock_filter net_perf_filter[] = {
 	return notify_fd;
 }
 
-void set_seccomp_response_to_continue(struct seccomp_notif *req,
-				      struct seccomp_notif_resp *resp)
-{
-	resp->id = req->id;
-	resp->val = 0;
-	resp->flags = SECCOMP_USER_NOTIF_FLAG_CONTINUE;
-	resp->error = 0;
-}
-
-void respond_to_seccomp_client(int notify_fd, struct seccomp_notif_resp *resp)
+static int child_send_fd(void *arg)
 {
-	if (ioctl(notify_fd, SECCOMP_IOCTL_NOTIF_SEND, resp) == -1) {
-		if (errno == ENOENT) {
-			printf(
-				"response failed with ENOENT; perhaps target "
-				"process's syscall was interrupted by signal?");
-		} else {
-			perror("ioctl-SECCOMP_IOCTL_NOTIF_SEND");
-		}
+	tsa_fds *fds = (tsa_fds *)arg;
+	PRINT_INFO("In child_send_fd : notify_fd = %d, going to wait\n", 
+		fds->notify_fd);
+	pthread_mutex_lock(&wait_to_send);
+	while(fds->notify_fd == -1) {
+        pthread_cond_wait(&ready_to_send, &wait_to_send);
+    }
+
+	if (send_fd(fds->fd_to_tsa, fds->notify_fd) == -1) {
+		PRINT_WARNING(
+			"Couldn't send fd to the TSA %s", strerror(errno));
+		return -1;
+	} else {
+		PRINT_INFO(
+			"Sent the notify fd to the seccomp agent socket to TSA");
 	}
-}
 
-void * catch_send_fd(void *fd_arg)
-{
-	struct seccomp_notif *req;
-	struct seccomp_notif_resp *resp;
-	struct seccomp_notif_sizes sizes;
-	struct epoll_event ev;
-	struct epoll_event evlist[MAX_EVENTS];
-	int epfd = -1, ready = 0;
-	int set_break = 0;
-	int notify_fd = *(int *)fd_arg;
+	pthread_mutex_unlock(&wait_to_send);
+	return 0;
+}
 
-	if (seccomp(SECCOMP_GET_NOTIF_SIZES, 0, &sizes) == -1) {
-        PRINT_WARNING("seccomp-SECCOMP_GET_NOTIF_SIZES not available?");
-		return NULL;
-	}
+void maybe_setup_seccomp_notifer() {
+	char *socket_path;
+	const int STACK_SIZE = 64 * 1024;
+	char *stack;
+	char *stack_top;
+	int notify_fd = -1;
+	int child_pid = -1;
 	
-	req = malloc(sizes.seccomp_notif);
-	if (req == NULL) {
-		PRINT_WARNING("malloc req error!");
-		return NULL;
+	tsa_fds *fds = malloc(sizeof(tsa_fds));
+	if (fds == NULL) {
+		PRINT_WARNING("Could not allocate fds");
+		return;
 	}
-
-	resp = malloc(sizes.seccomp_notif_resp);
-	if (resp == NULL) {
-		PRINT_WARNING("malloc resp error!");
-		return NULL;
+	fds->notify_fd = -1;
+	fds->fd_to_tsa = -1;
+	
+	stack = malloc(STACK_SIZE);
+	if (stack == NULL) {
+		PRINT_WARNING("Failure to setup child for clone");
+		return;
 	}
-
-	epfd = epoll_create(EPOLL_IGNORED_VAL);
-	if (epfd == -1) {
-		PRINT_WARNING("epoll_create error");
-		return NULL;
+	stack_top = stack + STACK_SIZE;
+	if (CHILD_SIG != 0 && CHILD_SIG != SIGCHLD) {
```

Comment:
Why do we ignore that here? Or setup an ignore for the child signal?

Reply:
The intention is to ignore everything that isn't a SIGCHILD as the other signals might be to terminate the process. We don't ignore SIGCHILD so that we can wait in the parent.